File: ios/SwiftKotlination/SwiftKotlination/Sources/Features/TopStories/Model/TopStoriesManager.swift
import Foundation

protocol TopStoriesManagerProtocol {
    @discardableResult
    func stories(_ observer: @escaping Observer<[Story]>) -> Disposable
    func fetchStories()
}

struct TopStoriesManager {
    let networkManager: NetworkManagerProtocol
}

// MARK: - Protocol Methods

extension TopStoriesManager: TopStoriesManagerProtocol {
    @discardableResult
    func stories(_ observer: @escaping Observer<[Story]>) -> Disposable {
        return networkManager
            .observe(.fetchTopStories) { result in
                switch result {
                case .success(let data):
                    do {
                        let topStories = try JSONDecoder().decode(TopStories.self, from: data)
                        observer(.success(topStories.results))
                    } catch {
                        observer(.failure(error))
                    }

                case .failure(let error):
                    observer(.failure(error))
                }
        }
    }

    func fetchStories() {
        networkManager.execute(.fetchTopStories)
    }
}

File: android/SwiftKotlination/app/src/main/kotlin/fr/jhandguy/swiftkotlination/features/topstories/model/TopStoriesManager.kt
package fr.jhandguy.swiftkotlination.features.topstories.model

import fr.jhandguy.swiftkotlination.network.NetworkManagerInterface
import fr.jhandguy.swiftkotlination.network.Request
import fr.jhandguy.swiftkotlination.observer.Disposable
import fr.jhandguy.swiftkotlination.observer.Observer
import fr.jhandguy.swiftkotlination.observer.Result
import kotlinx.serialization.UnstableDefault
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration.Companion.Stable

interface TopStoriesManagerInterface {
    suspend fun topStories(observer: Observer<TopStories>): Disposable
    suspend fun fetchStories()
}

class TopStoriesManager(private val networkManager: NetworkManagerInterface) : TopStoriesManagerInterface {
    @UnstableDefault
    override suspend fun topStories(observer: Observer<TopStories>) =
            networkManager.observe(Request.FetchTopStories) { result ->
                when (result) {
                    is Result.Success -> observer(Result.Success(Json(Stable.copy(strictMode = false)).parse(TopStories.serializer(), String(result.data))))
                    is Result.Failure -> observer(result)
                }
            }

    override suspend fun fetchStories() = networkManager.execute(Request.FetchTopStories)
}
